x/exchange/keeper/market.go

```diff
@@ -0,0 +1,1303 @@
+// isn't found or if there aren't any changes provided.
+func (k Keeper) UpdateMarketDetails(ctx sdk.Context, marketID uint32, marketDetails exchange.MarketDetails, updatedBy sdk.AccAddress) error {
+	if err := marketDetails.Validate(); err != nil {
+		return err
+	}
+
+	marketAcc := k.GetMarketAccount(ctx, marketID)
+	if marketAcc == nil {
+		return fmt.Errorf("market %d account not found", marketID)
+	}
+
+	if marketAcc.MarketDetails.Equal(marketDetails) {
+		return errors.New("no changes")
+	}
+
+	marketAcc.MarketDetails = marketDetails
+	k.accountKeeper.SetAccount(ctx, marketAcc)
+	k.emitEvent(ctx, exchange.NewEventMarketDetailsUpdated(marketID, updatedBy))
+
+	return nil
+}
+
+// storeMarket writes all the market fields to the state store (except MarketDetails which are in the account).
+func storeMarket(store sdk.KVStore, market exchange.Market) {
+	marketID := market.MarketId
+	setMarketKnown(store, marketID)
+	setCreateAskFlatFees(store, marketID, market.FeeCreateAskFlat)
+	setCreateBidFlatFees(store, marketID, market.FeeCreateBidFlat)
+	setSellerSettlementFlatFees(store, marketID, market.FeeSellerSettlementFlat)
+	setSellerSettlementRatios(store, marketID, market.FeeSellerSettlementRatios)
+	setBuyerSettlementFlatFees(store, marketID, market.FeeBuyerSettlementFlat)
+	setBuyerSettlementRatios(store, marketID, market.FeeBuyerSettlementRatios)
+	setMarketActive(store, marketID, market.AcceptingOrders)
+	setUserSettlementAllowed(store, marketID, market.AllowUserSettlement)
+	setAccessGrants(store, marketID, market.AccessGrants)
+	setReqAttrsAsk(store, marketID, market.ReqAttrCreateAsk)
+	setReqAttrsBid(store, marketID, market.ReqAttrCreateBid)
+}
+
+// initMarket is similar to CreateMarket but assumes the market has already been
+// validated and also allows for the market account to already exist.
+func (k Keeper) initMarket(ctx sdk.Context, store sdk.KVStore, market exchange.Market) {
+	if market.MarketId == 0 {
+		market.MarketId = nextMarketID(store)
+	}
+	marketID := market.MarketId
+
+	marketAddr := exchange.GetMarketAddress(marketID)
+	marketAcc := k.getMarketAccountByAddr(ctx, marketAddr)
+	if marketAcc != nil {
+		if !market.MarketDetails.Equal(marketAcc.MarketDetails) {
+			marketAcc.MarketDetails = market.MarketDetails
+			k.accountKeeper.SetAccount(ctx, marketAcc)
+		}
+	} else {
+		marketAcc = &exchange.MarketAccount{
+			BaseAccount:   &authtypes.BaseAccount{Address: marketAddr.String()},
+			MarketId:      marketID,
+			MarketDetails: market.MarketDetails,
+		}
+		k.accountKeeper.NewAccount(ctx, marketAcc)
+		k.accountKeeper.SetAccount(ctx, marketAcc)
+	}
+
+	storeMarket(store, market)
+}
+
+// CreateMarket saves a new market to the store with all the info provided.
+// If the marketId is zero, the next available one will be used.
+func (k Keeper) CreateMarket(ctx sdk.Context, market exchange.Market) (uint32, error) {
+	// Note: The Market is passed in by value, so any alterations to it here will be lost upon return.
```

Comment:
It's there to prevent someone from changing the arg to a `*exchange.Market` without realizing theres stuff in there depending on it not being a reference.

Reply:
Hm, if that's a concern then would you want to change that param to a pointer and manually make a copy of it? The code would speak for itself and we wouldn't have the concern of someone changing that param type.

Reply:
No. That's a whole lot of extra code just to get rid of a one line comment. It'd also be significantly less efficient than the automatic way that go does it. The code actually does speak to itself as long as you're familiar with some golang intricacies. This comment is a defensive measure to hopefully prevent someone less familiar with golang from unknowingly making a mistake.

Reply:
I usually just put those things into tests.  If a developer changes something and it breaks then mistake is apparent.

Reply:
I guess I'm not really understanding the problem because it should be the same amount of lines just handling both our concerns. I'm pretty sure go does pass by value and doesn't do a deep copy, which would be equivalent to this... We don't need the comment, and we aren't concerned about the user changing the param type.

```
func (k Keeper) CreateMarket(ctx sdk.Context, *m exchange.Market) (uint32, error) {
	market := *m
```

Reply:
This is exactly what comments are made for. It's explaining something that might not be obvious.

Everywhere that calls `CreateMarket` already has it as a value (not a reference), so I'd then have to update those calls to send in the reference to it. And then in here, I'd have to make sure it's not nil before doing anything with it, and also need an extra unit tests for that case, and I'd still have to have the unit tests make sure it's not being changed.

This one comment saves lots of code.

When I write the unit tests on it, I'll probably add a check that it's not being changed, and at that time, I'll consider removing this comment. I'm expecting questions about, "Why are you making sure this doesn't change when you're passing it in by value?" though, since I've gotten things like that in the past when I unit tested similar situations.

Reply:
I'm not sure that I agree that comments are a replacement for error checks unless there is a very specific reason why that the error check cannot be there (performance). However, now that I have put more thought into this I believe that this comment is actually doing more harm than good. The param is a shallow copy so the contents of Market aren't necessarily safe because it has pointers. This can actually cause more confusion than passing in by pointer because with the pointer they could expect modifications to persist. I would recommend one of the following:

1. Keep Market a value, remove the comment, and don't modify the values on the param.
2. Create a read only interface that exposes the methods needed and use that as the param type, and remove the comment.

An example of modifying a pass by value struct: https://go.dev/play/p/8TEYj0qd0on

Reply:
It's not the comment that's replacing the nil-case error check, it's that it can't possibly be nil here, so it doesn't make sense to check if it's nil.

A read-only interface is yet even more code that would need to be written and tested just to replace two lines of comments.

I'll consider removing the comment once there's a unit test ensuring the provided market doesn't change during this call. But again, the comment is pointing out behavior that isn't obvious.